controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -99,11 +137,414 @@ func (r *StorageClassClaimReconciler) reconcileConsumerPhases(ctx context.Contex
 }
 
 func (r *StorageClassClaimReconciler) reconcileProviderPhases(ctx context.Context, storagecluster *v1.StorageCluster) (reconcile.Result, error) {
-	log := ctrllog.FromContext(ctx)
-	log.Info("Running StorageClassClaim controller in Converged/Provider Mode")
+	r.log.Info("Running StorageClassClaim controller in Converged/Provider Mode")
+
+	r.storageClassClaim.Status.Phase = string(v1alpha1.StorageClassClaimStateConfiguring)
+
+	for _, cephResourceSpec := range r.storageClassClaim.Status.CephResources {
+		r.cephResourcesByName[cephResourceSpec.Name] = cephResourceSpec
+	}
+
+	if r.storageClassClaim.GetDeletionTimestamp().IsZero() {
+		if !controllers.Contains(r.storageClassClaim.GetFinalizers(), storageClassClaimFinalizer) {
+			r.log.Info("Finalizer not found for StorageClassClaim. Adding finalizer.", "StorageClassClaim", klog.KRef(r.storageClassClaim.Namespace, r.storageClassClaim.Name))
+			r.storageClassClaim.ObjectMeta.Finalizers = append(r.storageClassClaim.ObjectMeta.Finalizers, storageClassClaimFinalizer)
+			if err := r.Client.Update(ctx, r.storageClassClaim); err != nil {
+				r.log.Error(err, "Failed to update StorageClassClaim with finalizer.", "StorageClassClaim", klog.KRef(r.storageClassClaim.Namespace, r.storageClassClaim.Name))
+				return reconcile.Result{}, err
+			}
+		}
+		if r.storageClassClaim.Spec.Type == "blockpool" {
+			if err := r.reconcileCephClientRBDProvisioner(); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			if err := r.reconcileCephClientRBDNode(); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			if err := r.reconcileCephBlockPool(storagecluster); err != nil {
+				return reconcile.Result{}, err
+			}
+		} else if r.storageClassClaim.Spec.Type == "sharedfilesystem" {
+			if err := r.reconcileCephClientCephFSProvisioner(); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			if err := r.reconcileCephClientCephFSNode(); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			if err := r.reconcileCephFilesystemSubVolumeGroup(); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+		cephResourcesReady := true
+		for _, cephResource := range r.storageClassClaim.Status.CephResources {
+			if cephResource.Phase != "Ready" {
+				cephResourcesReady = false
+				break
+			}
+		}
+
+		if cephResourcesReady {
+			r.storageClassClaim.Status.Phase = string(v1alpha1.StorageClassClaimStateReady)
+		}
+
+	} else {
+		// todo deletion of the Ceph Resources
+	}
+
+	r.log.Info("Running StorageClassClaim controller in Converged/Provider Mode")
 	return reconcile.Result{}, nil
 }
 
+func (r *StorageClassClaimReconciler) initReconciler(request reconcile.Request) {
+	r.namespace = request.Namespace
+	ownerRef := metav1.OwnerReference{
+		UID:        r.storageClassClaim.UID,
+		APIVersion: r.storageClassClaim.APIVersion,
+		Kind:       r.storageClassClaim.Kind,
+		Name:       r.storageClassClaim.Name,
+	}
+	if r.storageClassClaim.Spec.Type == "cephBlockPool" {
```

Comment:
I initialized the storageClassClaimm inside initReconcile

so you are suggesting to initialize storageCluster inside initReconcile method but it will require the name for the resources, right?

Reply:
Ok, I see your point, so specifically for storagecluster we can skip the init here.